import { ValidationError } from './common';
import { validateClass } from './serverCommon';
import {
  schema,
  InternalServerError,
  {{#exceptions}}
  {{name}},
  {{/exceptions}}
  {{#classes}}
  {{name}},
  {{/classes}}
  {{#enums}}
  {{name}},
  {{/enums}}
  {{#bypassTypes}}
  {{name}},
  {{/bypassTypes}}
} from './interfaces';

{{#globals}}
{{#serverOnlyContext}}

export { ServerOnlyContext };
{{/serverOnlyContext}}
{{#serverContext}}
export type Context = {{{serverContext}}};
{{/serverContext}}
{{/globals}}
{{#classes}}
{{^attributes}}

{{#serverOnlyContext}}
export type {{{name}}}ContextExtractor = () => Promise<{{{serverOnlyContext}}}>;
{{/serverOnlyContext}}

const {{{name}}}Validator = validateClass(schema, '{{{name}}}');

export interface {{{name}}}Handler {
  {{#methods}}
  {{{name}}}({{#serverContext}}ctx: {{{serverContext}}}, {{/serverContext}}{{#parameters}}{{{name}}}{{#optional}}?{{/optional}}: {{{type}}}{{^last}}, {{/last}}{{/parameters}}): Promise<{{{returnType}}}>;
  {{/methods}}
}
{{#methods}}

export async function exec{{{className}}}{{{name}}}(body: any, fn: {{{className}}}Handler['{{{name}}}']{{#serverOnlyContext}}, extractContext: {{{className}}}ContextExtractor{{/serverOnlyContext}}, stackTraceInError: boolean = false) {
  let context: {{#serverContext}}{{{serverContext}}}{{/serverContext}}{{^serverContext}}{}{{/serverContext}} | undefined;
  try {
    {{{className}}}Validator('{{{name}}}', body);
    const { ctx: clientContextFromBody, ...args } = body;

    {{#clientContext}}
    const clientContext = clientContextFromBody as {{clientContext}};
    {{/clientContext}}
    {{^clientContext}}
    const clientContext = {};
    {{/clientContext}}
    {{#serverOnlyContext}}
    const serverOnlyContext = await extractContext();
    {{/serverOnlyContext}}
    {{^serverOnlyContext}}
    const serverOnlyContext = {};
    {{/serverOnlyContext}}
    context = { ...clientContext, ...serverOnlyContext };
    {{#serverContext}}
    return {
      status: 200,
      body: await fn(context{{#parameters}}, args.{{{name}}}{{/parameters}}),
      context,
    };
    {{/serverContext}}
    {{^serverContext}}
    return {
      status: 200,
      body: await fn({{#parameters}}args.{{{name}}}{{^last}}, {{/last}}{{/parameters}}),
      context,
    };
    {{/serverContext}}
  } catch (error) {
    const { stack, message, ...rest } = error;
    const processedError = stackTraceInError ? { stack: stack ? stack.toString() : undefined, ...rest } : rest;
    if (error instanceof ValidationError) {
      return {
        status: 400,
        error,
        body: {
          name: 'ValidationError',
          message,
          errors: error.errors,
        },
        context,
      };
    }
    {{#throws}}
    if (error instanceof {{{.}}}) {
      return {
        status: 500,
        error,
        body: {
          ...processedError,
          message,
          name: '{{{.}}}',
        },
        context,
      };
    }
    {{/throws}}
    return {
      status: 500,
      error,
      body: {
        ...processedError,
        message,
        name: 'InternalServerError',
      },
      context,
    };
  }
}
{{/methods}}
{{/attributes}}
{{/classes}}
